README.md
# egg-onerror

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![David deps][david-image]][david-url]
[![Known Vulnerabilities][snyk-image]][snyk-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/egg-onerror.svg?style=flat-square
[npm-url]: https://npmjs.org/package/egg-onerror
[travis-image]: https://img.shields.io/travis/eggjs/egg-onerror.svg?style=flat-square
[travis-url]: https://travis-ci.org/eggjs/egg-onerror
[codecov-image]: https://codecov.io/github/eggjs/egg-onerror/coverage.svg?branch=master
[codecov-url]: https://codecov.io/github/eggjs/egg-onerror?branch=master
[david-image]: https://img.shields.io/david/eggjs/egg-onerror.svg?style=flat-square
[david-url]: https://david-dm.org/eggjs/egg-onerror
[snyk-image]: https://snyk.io/test/npm/egg-onerror/badge.svg?style=flat-square
[snyk-url]: https://snyk.io/test/npm/egg-onerror
[download-image]: https://img.shields.io/npm/dm/egg-onerror.svg?style=flat-square
[download-url]: https://npmjs.org/package/egg-onerror

Default error handling plugin for egg.

## Install

```bash
$ npm i egg-onerror
```

## Usage

`egg-onerror` is on by default in egg. But you still can configure its properties to fits your scenarios.

- `errorPageUrl: String or Function` - If user request html pages in production environment and unexpected error happened, it will redirect user to `errorPageUrl`.
- `accepts: Function` - detect user's request accpet `json` or `html`.
- `all: Function` - customize error handler, if `all` present, negotiation will be ignored.
- `html: Function` - customize html error handler.
- `text: Function` - customize text error handler.
- `json: Function` - customize json error handler.
- `jsonp: Function` - customize jsonp error handler.
- `isProd: Function` - customize logic to detect if app is running on production environment. [Example](./test/fixtures/onerror-custom-isProd/config/config.default.js#L4)

```js
// config.default.js
// errorPageUrl support funtion
exports.onerror = {
  errorPageUrl: (err, ctx) => ctx.errorPageUrl || '/500',
};

// an accept detect function that mark all request with `x-requested-with=XMLHttpRequest` header accepts json.
function accepts(ctx) {
  if (ctx.get('x-requested-with') === 'XMLHttpRequest') return 'json';
  return 'html';
}
```

## Questions & Suggestions

Please open an issue [here](https://github.com/eggjs/egg/issues).

## License

[MIT](https://github.com/eggjs/egg-onerror/blob/master/LICENSE)


History.md

2.1.1 / 2022-08-18
==================

**fixes**
  * [[`00d2aa2`](http://github.com/eggjs/egg-onerror/commit/00d2aa2a073048dec9b9fc0fd0d868ecc0446830)] - fix: check file exists (#35) (吖猩 <<whxaxes@qq.com>>)

**others**
  * [[`54e12ba`](http://github.com/eggjs/egg-onerror/commit/54e12baa2eab9b47a2acd6ba0e6f6a0e55c92fc0)] - Create codeql-analysis.yml (fengmk2 <<fengmk2@gmail.com>>)
  * [[`55d27e6`](http://github.com/eggjs/egg-onerror/commit/55d27e60a9f1094e1a4555e82b47cab4799a57f8)] - chore: update travis (#31) (TZ | 天猪 <<atian25@qq.com>>)

2.1.0 / 2018-06-12
==================

**features**
  * [[`63cca2f`](http://github.com/eggjs/egg-onerror/commit/63cca2f3fb087583459e26e38c3874285b14aefd)] - feat: add replace template config (#28) (Harry Chen <<czy88840616@gmail.com>>)

2.0.0 / 2017-11-13
==================

**others**
  * [[`6861f8f`](http://github.com/eggjs/egg-onerror/commit/6861f8fb5df4a210afca2c7454dcca4ec1ccbae4)] - refactor: use async function and support egg@2 (#27) (Yiyu He <<dead_horse@qq.com>>)

1.6.0 / 2017-11-13
==================

**features**
  * [[`4a1b770`](http://github.com/eggjs/egg-onerror/commit/4a1b7707b28d3cc1e8bd69f4cca606305c507248)] - feat: support customize error handler (#26) (Yiyu He <<dead_horse@qq.com>>)
  * [[`37c06ce`](http://github.com/eggjs/egg-onerror/commit/37c06ce45fb671a3087f4e74aafcef1ac122360d)] - feat: support jsonp (#25) (Yiyu He <<dead_horse@qq.com>>)

**others**
  * [[`a0d4df2`](http://github.com/eggjs/egg-onerror/commit/a0d4df2830bf58903dd27e277f963e3d52d32587)] - chore: fix README & update deps & fix test (#23) (TZ | 天猪 <<atian25@qq.com>>)
  * [[`e49252e`](http://github.com/eggjs/egg-onerror/commit/e49252e3a648abbefc562635e163c4b9dd28e57d)] - docs: fix typo (#22) (TZ | 天猪 <<atian25@qq.com>>)

1.5.0 / 2017-07-20
==================

  * feat: errorPageUrl support function (#21)

1.4.6 / 2017-06-20
==================

  * fix: only output simple error html on unittest (#19)

1.4.5 / 2017-06-20
==================

  * fix: should show 4xx status error html (#18)

1.4.4 / 2017-06-12
==================

  * fix: make error style more good (#17)
  * fix: add error-title's line-height (#16)

1.4.3 / 2017-06-04
==================

  * docs: fix License url (#15)

1.4.2 / 2017-06-01
==================

  * fix: remove error detail on JSON response (#14)

1.4.1 / 2017-06-01
==================

  * fix: add missing files on package.json (#13)

1.4.0 / 2017-05-31
==================

  * feat: better error page on development (#12)

1.3.0 / 2017-01-22
==================

  * feat: use ctx.acceptJSON (#11)

1.2.2 / 2017-01-13
==================

  * fix: should keep support `*.json` ext to detect response type (#10)

1.2.1 / 2017-01-13
==================

  * fix: add agent.js to package.files (#9)

1.2.0 / 2017-01-13
==================

  * feat: support options.accepts
  * refactor: remove accpetJSON

1.1.0 / 2016-11-09
==================

  * feat: should watch error event on agent (#7)

1.0.0 / 2016-10-21
==================

  * deps: upgrade koa-onerror (#6)
  * fix: make sure ctx always exists (#3)

0.0.3 / 2016-07-16
==================

  * fix: fix this is undefined on arrow function (#1)

0.0.2 / 2016-07-13
==================
  * init code


test/onerror.test.js
'use strict';

const fs = require('fs');
const pedding = require('pedding');
const mm = require('egg-mock');
const rimraf = require('rimraf');
const path = require('path');
const assert = require('assert');

describe('test/onerror.test.js', () => {
  let app;
  before(() => {
    mm.env('local');
    mm.consoleLevel('NONE');
    app = mm.app({ baseDir: 'onerror' });
    return app.ready();
  });
  after(() => app.close());

  afterEach(mm.restore);

  it('should handle error not in the req/res cycle with no ctx', function* () {
    mm.consoleLevel('NONE');
    const app = mm.app({
      baseDir: 'mock-test-error',
    });
    yield app.ready();
    const err = new Error('mock test error');
    app.emit('error', err, null);
    err.status = 400;
    app.emit('error', err, null);
    app.close();
  });

  it('should handle status:-1 as status:500', () => {
    return app.httpRequest()
      .get('/?status=-1')
      .expect(/<h1 class="box">Error in &#x2F;\?status&#x3D;-1<\/h1>/)
      .expect(500);
  });

  it('should handle status:undefined as status:500', () => {
    return app.httpRequest()
      .get('/')
      .expect(/<div class="context">test error<\/div>/)
      .expect(500);
  });

  it('should handle not exists file in stack without error', () => {
    return app.httpRequest()
      .get('/unknownFile')
      .expect(/<div class="context">test error<\/div>/)
      .expect(500);
  });

  it('should handle escape xss', () => {
    return app.httpRequest()
      .get('/?message=<script></script>')
      .expect(/&lt;script&gt;&lt;&#x2F;script&gt;/)
      .expect(500);
  });

  it('should handle status:1 as status:500', () => {
    return app.httpRequest()
      .get('/?status=1')
      .expect(/<div class="context">test error<\/div>/)
      .expect(500);
  });

  it('should handle status:400', () => {
    return app.httpRequest()
      .get('/?status=400')
      .expect(/<div class="context">test error<\/div>/)
      .expect(400);
  });

  it('should return error json format when Accept is json', () => {
    return app.httpRequest()
      .get('/user')
      .set('Accept', 'application/json')
      .expect(res => {
        assert(res.body);
        assert(res.body.message === 'test error');
        assert(res.body.stack.includes('Error: test error'));
        // should not includes error detail
        assert(!res.body.frames);
      })
      .expect(500);
  });

  it('should return error json format when request path match *.json', () => {
    return app.httpRequest()
      .get('/user.json')
      .expect(res => {
        assert(res.body);
        assert(res.body.message === 'test error');
        assert(res.body.stack.includes('Error: test error'));
        assert(res.body.status === 500);
        assert(res.body.name === 'Error');
      })
      .expect(500);
  });

  it('should support custom accpets return err.stack', () => {
    mm(app.config.onerror, 'accepts', ctx => {
      if (ctx.get('x-requested-with') === 'XMLHttpRequest') return 'json';
      return 'html';
    });
    return app.httpRequest()
      .get('/user.json')
      .set('x-requested-with', 'XMLHttpRequest')
      .expect(/"message":"test error"/)
      .expect(/"stack":/)
      .expect(500);
  });

  it('should return err.stack when unittest', () => {
    mm(app.config, 'env', 'unittest');
    return app.httpRequest()
      .get('/user.json')
      .set('Accept', 'application/json')
      .expect(/"message":"test error"/)
      .expect(/"stack":/)
      .expect(500);
  });

  it('should return err status message', () => {
    mm(app.config, 'env', 'prod');
    return app.httpRequest()
      .get('/user.json')
      .set('Accept', 'application/json')
      .expect({ message: 'Internal Server Error' })
      .expect(500);
  });

  it('should return err.errors', () => {
    return app.httpRequest()
      .get('/user.json?status=400&errors=test')
      .set('Accept', 'application/json')
      .expect(/test/)
      .expect(400);
  });

  it('should return err json at prod env', () => {
    mm(app.config, 'env', 'prod');
    return app.httpRequest()
      .get('/user.json?status=400&errors=test')
      .set('Accept', 'application/json')
      .expect(/test/)
      .expect({
        errors: 'test',
        message: 'test error',
      })
      .expect('Content-Type', 'application/json; charset=utf-8')
      .expect(400);
  });

  it('should return 4xx html at prod env', () => {
    mm(app.config, 'env', 'prod');
    return app.httpRequest()
      .post('/test?status=400&errors=test')
      .set('Accept', 'text/html')
      .expect('<h2>400 Bad Request</h2>')
      .expect('Content-Type', 'text/html; charset=utf-8')
      .expect(400);
  });

  it('should return 500 html at prod env', () => {
    mm(app.config, 'env', 'prod');
    mm(app.config.onerror, 'errorPageUrl', '');
    return app.httpRequest()
      .post('/test?status=502&errors=test')
      .set('Accept', 'text/html')
      .expect('<h2>Internal Server Error, real status: 502</h2>')
      .expect('Content-Type', 'text/html; charset=utf-8')
      .expect(500);
  });

  it('should return err json at non prod env', () => {
    return app.httpRequest()
      .get('/user.json?status=400&errors=test')
      .set('Accept', 'application/json')
      .expect(/test/)
      .expect({
        errors: 'test',
        message: 'test error',
      })
      .expect(400);
  });

  it('should return parsing json error on html response', () => {
    return app.httpRequest()
      .post('/test?status=400')
      .send({ test: 1 })
      .set('Content-Type', 'application/json')
      .expect(/Problems parsing JSON/)
      .expect('Content-Type', 'text/html; charset=utf-8')
      .expect(400);
  });

  it('should return parsing json error on json response', () => {
    return app.httpRequest()
      .post('/test?status=400')
      .send({ test: 1 })
      .set('Content-Type', 'application/json')
      .set('Accept', 'application/json')
      .expect({
        message: 'Problems parsing JSON',
      })
      .expect('Content-Type', 'application/json; charset=utf-8')
      .expect(400);
  });

  it('should redirect to error page', () => {
    mm(app.config, 'env', 'test');
    return app.httpRequest()
      .get('/?status=500')
      .expect('Location', 'https://eggjs.com/500.html?real_status=500')
      .expect(302);
  });

  it('should handle 403 err', () => {
    mm(app.config, 'env', 'prod');
    return app.httpRequest()
      .get('/?status=403&code=3')
      .expect('<h2>403 Forbidden</h2>')
      .expect(403);
  });

  it('should return jsonp style', () => {
    mm(app.config, 'env', 'prod');
    return app.httpRequest()
      .get('/jsonp?callback=fn')
      .expect('content-type', 'application/javascript; charset=utf-8')
      .expect('/**/ typeof fn === \'function\' && fn({"message":"Internal Server Error"});')
      .expect(500);
  });

  describe('customize', () => {
    let app;
    before(() => {
      mm.consoleLevel('NONE');
      app = mm.app({
        baseDir: 'onerror-customize',
      });
      return app.ready();
    });
    after(() => app.close());

    it('should support customize json style', () => {
      mm(app.config, 'env', 'prod');
      return app.httpRequest()
        .get('/user.json')
        .expect('content-type', 'application/json; charset=utf-8')
        .expect({ msg: 'error' })
        .expect(500);
    });

    it('should return jsonp style', () => {
      mm(app.config, 'env', 'prod');
      return app.httpRequest()
        .get('/jsonp?callback=fn')
        .expect('content-type', 'application/javascript; charset=utf-8')
        .expect('/**/ typeof fn === \'function\' && fn({"msg":"error"});')
        .expect(500);
    });

    it('should handle html by default', () => {
      mm(app.config, 'env', 'test');
      return app.httpRequest()
        .get('/?status=500')
        .expect('Location', 'https://eggjs.com/500.html?real_status=500')
        .expect(302);
    });
  });

  if (process.platform !== 'win32') {
    it('should log warn 4xx', function* () {
      rimraf.sync(path.join(__dirname, 'fixtrues/onerror-4xx/logs'));
      const app = mm.app({
        baseDir: 'onerror-4xx',
      });
      yield app.ready();
      yield app.httpRequest()
        .post('/body_parser')
        .set('Content-Type', 'application/x-www-form-urlencoded')
        .send({ foo: new Buffer(1024 * 1000).fill(1).toString() })
        .expect(/request entity too large/)
        .expect(413);
      yield app.close();

      const warnLog = path.join(__dirname, 'fixtures/onerror-4xx/logs/onerror-4xx/onerror-4xx-web.log');
      assert(/POST \/body_parser] nodejs\..*?Error: request entity too large/.test(fs.readFileSync(warnLog, 'utf8')));
    });
  }

  describe('no errorpage', () => {
    let app;
    before(() => {
      mm.consoleLevel('NONE');
      app = app = mm.app({
        baseDir: 'onerror-no-errorpage',
      });
      return app.ready();
    });
    after(() => app.close());

    it('should display 500 Internal Server Error', () => {
      mm(app.config, 'env', 'prod');
      return app.httpRequest()
        .get('/?status=500')
        .expect(500)
        .expect(/Internal Server Error, real status: 500/);
    });
  });

  describe('app.errorpage.url=/500', () => {
    let app;
    before(() => {
      mm.consoleLevel('NONE');
      app = app = mm.app({
        baseDir: 'onerror-custom-500',
      });
      return app.ready();
    });
    after(() => app.close());

    it('should redirect to error page', function* () {
      mm(app.config, 'env', 'prod');

      yield app.httpRequest()
        .get('/mockerror')
        .expect('Location', '/500?real_status=500')
        .expect(302);

      yield app.httpRequest()
        .get('/mock4xx')
        .expect('<h2>400 Bad Request</h2>')
        .expect(400);

      yield app.httpRequest()
        .get('/500')
        .expect('hi, this custom 500 page')
        .expect(500);

      yield app.httpRequest()
        .get('/special')
        .expect('Location', '/specialerror?real_status=500')
        .expect(302);
    });
  });

  describe('onerror.ctx.error env=local', () => {
    let app;
    before(() => {
      mm.env('local');
      mm.consoleLevel('NONE');
      app = mm.app({
        baseDir: 'onerror-ctx-error',
      });
      return app.ready();
    });
    after(() => app.close());

    it('should 500 full html', () => {
      return app.httpRequest()
        .get('/error')
        .expect(500)
        .expect(/you can&#x60;t get userId\./);
    });
  });

  describe('onerror.ctx.error env=unittest', () => {
    let app;
    before(() => {
      mm.consoleLevel('NONE');
      app = mm.app({
        baseDir: 'onerror-ctx-error',
      });
      return app.ready();
    });
    after(() => app.close());

    it('should 500 simple html', () => {
      return app.httpRequest()
        .get('/error')
        .expect(500)
        .expect(/you can`t get userId\./);
    });
  });

  describe('appErrorFilter', () => {
    let app;
    before(() => {
      mm.consoleLevel('NONE');
      app = mm.app({
        baseDir: 'custom-listener-onerror',
      });
      return app.ready();
    });
    after(() => app.close());

    it('should ignore error log', () => {
      mm(app.logger, 'log', () => {
        throw new Error('should not excute');
      });

      return app.httpRequest()
        .get('/?name=IgnoreError')
        .expect(500);
    });

    it('should custom log error log', done => {
      done = pedding(done, 2);
      mm(app.logger, 'error', err => {
        assert.equal(err, 'error happened');
        done();
      });

      app.httpRequest()
        .get('/?name=CustomError')
        .expect(500)
        .then(() => done(), done);
    });

    it('should default log error', done => {
      done = pedding(done, 2);
      mm(app.logger, 'log', (LEVEL, args) => {
        assert.equal(args[0].name, 'OtherError');
        done();
      });

      app.httpRequest()
        .get('/?name=OtherError')
        .expect(500)
        .then(() => done(), done);
    });
  });

  describe('agent emit error', () => {
    let app;
    before(() => {
      app = mm.cluster({
        baseDir: 'agent-error',
      });
      app.debug();
      return app.ready();
    });
    after(() => app.close());

    it('should log error', function* () {
      console.log('app.stderr: %j', app.stderr);
      console.log(app.stdout);
      yield app.close();
      assert(/nodejs.Error: emit error/.test(app.stderr));
    });
  });

  describe('replace onerror default template', () => {

    let app = null;
    before(() => {
      mm.consoleLevel('NONE');
      app = mm.app({
        baseDir: 'onerror-custom-template',
      });
      return app.ready();
    });

    after(() => app.close());

    afterEach(mm.restore);

    it('should use custom template', () => {
      mm(app.config, 'env', 'local');
      return app.httpRequest()
        .get('/')
        .expect(/custom template/)
        .expect(500);
    });

  });
});


test/error_handler.test.js
'use strict';

const mm = require('egg-mock');
const assert = require('assert');

describe('test/error_handler.test.js', () => {
  describe('default error handler', () => {
    let app;
    before(() => {
      app = mm.app({
        baseDir: 'any-error-handler',
      });
      return app.ready();
    });
    after(() => app.close());

    describe('error', () => {
      it('should handler error', async () => {
        await app.httpRequest()
          .get('/error')
          .set('Accept', 'application/json')
          .expect({
            code: 'UNKNOWN_ERROR',
            message: 'any error',
          })
          .expect(500);

        await app.httpRequest()
          .get('/error')
          .set('Accept', 'text/plain')
          .expect({
            code: 'UNKNOWN_ERROR',
            message: 'any error',
          })
          .expect(500);

        await app.httpRequest()
          .get('/error')
          .set('Accept', 'text/html')
          .expect({
            code: 'UNKNOWN_ERROR',
            message: 'any error',
          })
          .expect(500);

        await app.httpRequest()
          .get('/error')
          .expect({
            code: 'UNKNOWN_ERROR',
            message: 'any error',
          })
          .expect(500);
      });

    });

    describe('egg error', () => {
      it('should handler egg error', async () => {
        await app.httpRequest()
          .get('/egg-error')
          .set('Accept', 'application/json')
          .expect({
            code: 'CUSTOM_ERROR',
            message: 'any egg error',
          })
          .expect(422);

        await app.httpRequest()
          .get('/egg-error')
          .set('Accept', 'text/plain')
          .expect({
            code: 'CUSTOM_ERROR',
            message: 'any egg error',
          })
          .expect(422);

        await app.httpRequest()
          .get('/egg-error')
          .set('Accept', 'text/html')
          .expect({
            code: 'CUSTOM_ERROR',
            message: 'any egg error',
          })
          .expect(422);

        await app.httpRequest()
          .get('/egg-error')
          .expect({
            code: 'CUSTOM_ERROR',
            message: 'any egg error',
          })
          .expect(422);
      });
    });

    describe('egg exception', () => {
      it('should handler egg exception', async () => {
        await app.httpRequest()
          .get('/egg-exception')
          .set('Accept', 'application/json')
          .expect({
            code: 'CUSTOM_EXCEPTION',
            message: 'any egg exception',
          })
          .expect(500);

        await app.httpRequest()
          .get('/egg-exception')
          .set('Accept', 'text/plain')
          .expect({
            code: 'CUSTOM_EXCEPTION',
            message: 'any egg exception',
          })
          .expect(500);

        await app.httpRequest()
          .get('/egg-exception')
          .set('Accept', 'text/html')
          .expect({
            code: 'CUSTOM_EXCEPTION',
            message: 'any egg exception',
          })
          .expect(500);

        await app.httpRequest()
          .get('/egg-exception')
          .expect({
            code: 'CUSTOM_EXCEPTION',
            message: 'any egg exception',
          })
          .expect(500);
      });
    });

  });

  describe('custom error handler', () => {
    let app;
    before(() => {
      app = mm.app({
        baseDir: 'custom-error-handler',
      });
      return app.ready();
    });
    after(() => app.close());

    describe('error', () => {
      it('should handler error when accept json', async () => {
        await app.httpRequest()
          .get('/error')
          .set('Accept', 'application/json')
          .expect({
            code: 'UNKNOWN_ERROR',
            message: 'custom error',
          })
          .expect(500);
      });

      it('should handler error when accept text', async () => {
        await app.httpRequest()
          .get('/error')
          .set('Accept', 'text/plain')
          .expect('custom error')
          .expect(500);
      });

      it('should handler error when accept html', async () => {
        await app.httpRequest()
          .get('/error')
          .set('Accept', 'text/html')
          .expect('<h2>UNKNOWN_ERROR</h2>\n<div>custom error</div>')
          .expect(500);
      });

      it('should handler error without accept', async () => {
        await app.httpRequest()
          .get('/error')
          .expect('<h2>UNKNOWN_ERROR</h2>\n<div>custom error</div>')
          .expect(500);
      });
    });

    describe('egg error', () => {
      it('should handler egg error when accept json', async () => {
        await app.httpRequest()
          .get('/egg-error')
          .set('Accept', 'application/json')
          .expect({
            code: 'CUSTOM_ERROR',
            message: 'custom egg error',
          })
          .expect(422);
      });

      it('should handler egg error when accept text', async () => {
        await app.httpRequest()
          .get('/egg-error')
          .set('Accept', 'text/plain')
          .expect('custom egg error')
          .expect(422);
      });

      it('should handler egg error when accept html', async () => {
        await app.httpRequest()
          .get('/egg-error')
          .set('Accept', 'text/html')
          .expect('<h2>CUSTOM_ERROR</h2>\n<div>custom egg error</div>')
          .expect(422);
      });

      it('should handler egg error without accept', async () => {
        await app.httpRequest()
          .get('/egg-error')
          .set('Accept', 'text/html')
          .expect('<h2>CUSTOM_ERROR</h2>\n<div>custom egg error</div>')
          .expect(422);
      });
    });

    describe('egg exception', () => {
      it('should handler egg exception when accept json', async () => {
        await app.httpRequest()
          .get('/egg-exception')
          .set('Accept', 'application/json')
          .expect({
            code: 'CUSTOM_EXCEPTION',
            message: 'custom egg exception',
          })
          .expect(500);
      });

      it('should handler egg exception when accept text', async () => {
        await app.httpRequest()
          .get('/egg-exception')
          .set('Accept', 'text/plain')
          .expect('custom egg exception')
          .expect(500);
      });

      it('should handler egg exception when accept html', async () => {
        await app.httpRequest()
          .get('/egg-exception')
          .set('Accept', 'text/html')
          .expect('<h2>CUSTOM_EXCEPTION</h2>\n<div>custom egg exception</div>')
          .expect(500);
      });

      it('should handler egg exception without accept', async () => {
        await app.httpRequest()
          .get('/egg-exception')
          .expect('<h2>CUSTOM_EXCEPTION</h2>\n<div>custom egg exception</div>')
          .expect(500);
      });
    });
  });

  describe('unmatch error handler', () => {
    let app;
    before(() => {
      app = mm.app({
        baseDir: 'unmatch-error-handler',
      });
      return app.ready();
    });
    after(() => app.close());

    it('should throw when accept json', async () => {
      const res = await app.httpRequest()
        .get('/error')
        .set('Accept', 'application/json')
        .expect(500);
      assert(res.body.code === 'UNKNOWN_ERROR');
      assert(res.body.message === 'error');
    });

    it('should throw when accept html', async () => {
      const res = await app.httpRequest()
        .get('/error')
        .set('Accept', 'text/html')
        .expect(500);

      assert(res.text.includes('UnknownError: error'));
    });
  });

});


test/fixtures/agent-error/agent.js
'use strict';

module.exports = agent => {
  const done = agent.readyCallback();
  setTimeout(() => {
    done(new Error('emit error'));
  }, 500);
};


app.js
'use strict';

const http = require('http');
const fs = require('fs');
const onerror = require('koa-onerror');
const ErrorView = require('./lib/error_view');
const {
  isProd,
  detectStatus,
  detectErrorMessage,
  accepts,
} = require('./lib/utils');

module.exports = app => {
  // logging error
  const config = app.config.onerror;
  const viewTemplate = fs.readFileSync(config.templatePath, 'utf8');

  app.on('error', (err, ctx) => {
    ctx = ctx || app.createAnonymousContext();
    if (config.appErrorFilter && !config.appErrorFilter(err, ctx)) return;

    const status = detectStatus(err);
    // 5xx
    if (status >= 500) {
      try {
        ctx.logger.error(err);
      } catch (ex) {
        app.logger.error(err);
        app.logger.error(ex);
      }
      return;
    }

    // 4xx
    try {
      ctx.logger.warn(err);
    } catch (ex) {
      app.logger.warn(err);
      app.logger.error(ex);
    }
  });

  const errorOptions = {
    // support customize accepts function
    accepts() {
      const fn = config.accepts || accepts;
      return fn(this);
    },

    html(err) {
      const status = detectStatus(err);
      const errorPageUrl = typeof config.errorPageUrl === 'function'
        ? config.errorPageUrl(err, this)
        : config.errorPageUrl;

      // If it's null, then use default isProd logic
      let displayErrors = config.displayErrors === null ? !isProd(app) : config.displayErrors;

      // It also can be a function
      if (typeof displayErrors === 'function') {
        displayErrors = displayErrors.call(app);
      }

      // keep the real response status
      this.realStatus = status;
      // don't respond any error message in production env
      // if displayErrors set false, or unset, then use default logic
      if (!displayErrors) {
        // 5xx
        if (status >= 500) {
          if (errorPageUrl) {
            const statusQuery =
              (errorPageUrl.indexOf('?') > 0 ? '&' : '?') +
              `real_status=${status}`;
            return this.redirect(errorPageUrl + statusQuery);
          }
          this.status = 500;
          this.body = `<h2>Internal Server Error, real status: ${status}</h2>`;
          return;
        }
        // 4xx
        this.status = status;
        this.body = `<h2>${status} ${http.STATUS_CODES[status]}</h2>`;
        return;
      }
      // show simple error format for unittest
      if (app.config.env === 'unittest') {
        this.status = status;
        this.body = `${err.name}: ${err.message}\n${err.stack}`;
        return;
      }

      const errorView = new ErrorView(this, err, viewTemplate);
      this.body = errorView.toHTML();
    },

    json(err) {
      const status = detectStatus(err);
      let errorJson = {};

      this.status = status;
      const code = err.code || err.type;
      const message = detectErrorMessage(this, err);

      if (isProd(app)) {
        // 5xx server side error
        if (status >= 500) {
          errorJson = {
            code,
            // don't respond any error message in production env
            message: http.STATUS_CODES[status],
          };
        } else {
          // 4xx client side error
          // addition `errors`
          errorJson = {
            code,
            message,
            errors: err.errors,
          };
        }
      } else {
        errorJson = {
          code,
          message,
          errors: err.errors,
        };

        if (status >= 500) {
          // provide detail error stack in local env
          errorJson.stack = err.stack;
          errorJson.name = err.name;
          for (const key in err) {
            if (!errorJson[key]) {
              errorJson[key] = err[key];
            }
          }
        }
      }

      this.body = errorJson;
    },

    js(err) {
      errorOptions.json.call(this, err, this);

      if (this.createJsonpBody) {
        this.createJsonpBody(this.body);
      }
    },
  };

  // support customize error response
  [ 'all', 'html', 'json', 'text', 'js' ].forEach(type => {
    if (config[type]) errorOptions[type] = config[type];
  });
  onerror(app, errorOptions);
};


config/config.default.js
'use strict';
const path = require('path');

exports.onerror = {
  // 5xx error will redirect to ${errorPageUrl}
  // won't redirect in local env
  errorPageUrl: '',
  // will excute `appErrorFilter` when emit an error in `app`
  // If `appErrorFilter` return false, egg-onerror won't log this error.
  // You can logging in `appErrorFilter` and return false to override the default error logging.
  appErrorFilter: null,
  // default template path
  templatePath: path.join(__dirname, '../lib/onerror_page.mustache'),
  // Set displayErrors to true, to display erros.
  // If it's not set, then use the default isProd logic for it.
  displayErrors: null,
  // Sections you want to show in error page.
  // Default to show all sections.
  displaySections: [ 'CodeFrames', 'Headers', 'Cookies', 'AppInfo' ],
};


app/middleware/egg_onerror_handler.js
'use strict';

const { EggError, ErrorType, InternalServerError } = require('egg-errors');
const accepts = require('accepts');

class UnknownError extends InternalServerError {
  constructor(message) {
    super(message);
    this.message = message;
    this.code = 'UNKNOWN_ERROR';
  }
}

module.exports = (_, app) => {
  const errorHandler = app.config.onerror.errorHandler;
  const acceptContentType = app.config.onerror.accepts ||
    ((ctx, ...args) => accepts(ctx.req).type(args));

  return async function bizErrorHandler(ctx, next) {
    try {
      await next();
    } catch (e) {
      let err = e;
      const errorType = EggError.getType(err);
      switch (errorType) {
        case ErrorType.ERROR: break;

        // if error is an EggError Exception, it will pring error log
        case ErrorType.EXCEPTION: {
          ctx.logger.error(err);
          // force set error status
          err.status = 500;
          break;
        }

        // if error is not recognized by EggError,
        // it will be converted to UnknownError
        case ErrorType.BUILTIN: {
          err = UnknownError.from(err);
          ctx.logger.error(err);
          break;
        }

        // getType not work
        default:
      }

      // handle the error
      const contentType = acceptContentType(ctx, 'html', 'text', 'json');

      if (contentType === 'json' && errorHandler.json) {
        await errorHandler.json(ctx, err);
      } else if (contentType === 'text' && errorHandler.text) {
        await errorHandler.text(ctx, err);
      } else if (contentType === 'html' && errorHandler.html) {
        await errorHandler.html(ctx, err);
      } else if (errorHandler.any) {
        await errorHandler.any(ctx, err);
      } else {
        throw err;
      }
    }
  };
};


app/middleware/onerror_biz_handler.js
'use strict';

const { EggError, ErrorType } = require('egg-errors');
const { accepts } = require('../../lib/utils');

module.exports = (_, app) => {
  const application = app.config.onerror.application;

  return async function onerrorBizHandler(ctx, next) {
    try {
      await next();
    } catch (err) {
      const fn = app.config.onerror.accepts || accepts;
      const contentType = fn(ctx, 'html', 'text', 'json');
      console.info(contentType);
      const type = EggError.getType(err);
      console.log(type);
      switch (type) {
        case ErrorType.ERROR: {
          ctx.status = err.status || 500;
          ctx.body = format(err, contentType);
          break;
        }

        case ErrorType.EXCEPTION: {
          ctx.logger.error(err);
          err.status = 500;
          ctx.body = format(err, contentType);
          break;
        }

        case ErrorType.BUILTIN:
        default:
          throw err;
      }
    }

    function format(err, contentType) {
      if (contentType === 'json' && application.formatJSON) {
        ctx.body = application.formatJSON(err);
      } else if (contentType === 'text' && application.formatText) {
        ctx.body = application.formatText(err);
      } else if (contentType === 'html' && application.formatHtml) {
        ctx.body = application.formatHtml(err);
      } else {
        throw err;
      }
    }
  };

};


lib/utils.js
'use strict';

exports.detectErrorMessage = function(ctx, err) {
  // detect json parse error
  if (err.status === 400 &&
    err.name === 'SyntaxError' &&
    ctx.request.is('application/json', 'application/vnd.api+json', 'application/csp-report')) {
    return 'Problems parsing JSON';
  }
  return err.message;
};

exports.detectStatus = function(err) {
  // detect status
  let status = err.status || 500;
  if (status < 200) {
    // invalid status consider as 500, like urllib will return -1 status
    status = 500;
  }
  return status;
};

exports.accepts = function(ctx) {
  if (ctx.acceptJSON) return 'json';
  if (ctx.acceptJSONP) return 'js';
  return 'html';
};

exports.isProd = function(app) {
  let isProd = function(app) { return app.config.env !== 'local' && app.config.env !== 'unittest'; };

  if (app.config.onerror && (typeof app.config.onerror.isProd === 'function')) {
    isProd = app.config.onerror.isProd;
  }

  return isProd(app);
};
